Add properties for date parsing

basememara opened the pull request:
Properties and functions added for parsing dates with unit tests.

For flexibility, you can let `JASON` parse the date using a default ISO date formatting string (`yyyy-MM-dd'T'HH:mm:ss`), or pass in a string format or `NSDateFormatter`:

```
let json: JSON = [
    "date1": "2016-04-12T13:29:32",
    "date2": "16/04/2016",
    "date3": "April 10, 2020"
]

json["date1"].dateValue
json["date2"].dateValue("dd-MM-yyyy")

let formatter = NSDateFormatter()
formatter.dateFormat = "MMMM dd, yyyy"

json["date3"].dateValue(formatter)
```

This is related to discussion #15.

delba commented:
Thanks for you PR @basememara !

I'm a bit mixed though as it isn't consistent with the other properties and for my part, I don't really mind writing:

``` swift
let formatter = NSDateFormatter("...")

let date = formatter.dateFromString(json["created_at"].stringValue)
```

Also a few things:
1. `<type>Value` don't return an optional
2. Prefix Foundation/CG types with ns/cg (cf existing properties)
3. Creating a date formatter for every date is slow. It's better to only ask for a formatter as a param.
4. The formatter should be passed when we create the `JSONKey`

basememara commented:
Hey @delba, thx for the feedback and for an awesome library! Below are my thoughts:

1) I don't know how I missed that, I updated it so it's `json["created_at"].date` instead for optionals consistency.

2) Done, so now it's `json["created_at"].nsDate` instead (is this what you meant?)

3) I've updated it so it's using a cached static array date formatters, with ISO-8601 added by default. If other date formats are needed, then a new date formatter is created/cached on the fly and never created again for performance.

4) Passing in a formatter is still an option, but for more convenience you can use ISO-8601 by default or simply pass in a string format which will only be created once ever by JASON or your app.

Also, I didn't mind doing the way your code snippet shows at first, but started to get really tedious. The same could be said for other types too, such as `Int(json["age"].string)`. Adding dates to the list of supported types makes it more complete since almost every other type is supported and everything started to feel more consistent.

I realize it adds more code and management for `JASON` so I guess we are weighing if the convenience outweighs it or if `JASON` can handle dates better than the end dev would in regards to redundancy, boilerplate, performance, etc.

Let me know what you think from here. Thx!

delba commented:
> The same could be said for other types too, such as Int(json["age"].string).

**Very** fair point ahah 😂  

What I really dislike about passing a format string every time we want access a property is that it's way to easy to mess up:

``` swift
json["created_at"].date(stringFormat: "<format>")

// later...

json["updated_at"].date(stringFormat: "<formatWithTypo>")
```

Passing a string format when using `JSONKey` removes of the benefits of this feature...

It's fair to assume that in 99% of the cases, we'll only use one date format.
So here is what we should do:

``` swift
extension JSON {
    public static var dateFormatter = NSDateFormatter()

    public var nsDate: NSDate? {
        return JSON.dateFormatter.dateFromString(stringValue)
    }
}

extension subscript(key: JSONKey<NSDate?>) -> NSDate? {
    return self[key.type].nsDate
}
```

**No default format please !!!**

If you want to change the format:

``` swift
JSON.dateFormatter.dateFormat = "..."

// or even

JSON.dateFormatter = myCustomDateFormatter
```

basememara commented:
Oooh that's nice! Short, sweet, and makes no assumptions, I love it!! I committed the update.

delba commented:
Awesome, thanks!

Just a few more things please:
- move `public static var dateFormatter` from `Properties.swift` to `JSON.swift` (above `public let object`) : `JSONKey.swift` also uses it... 
- rename the `MARK: - Date` to `MARK: - NSDate` and move these sections above `NSURL` (so that they'll be in the same order as in the README references section)
- add a very short note in the references section of the README on how to use it
- squash everything into 1 or 2 commits ahah

delba commented on README.md:
```diff
@@ -110,6 +110,17 @@ let personJSON = peopleJSON[personKey]
 let nameKey = JSONKey<String>("name")
 let name = personJSON[nameKey]
 
+// With a date key:
```
Could you move this into a shorter not in the "References" please?

basememara commented:
Ok I think I covered all your points, thanks for the tips. Let me know if you think of anything else.

basememara commented:
Squashing didn't condense as much as I hoped, but I think you can squash and merge PR's from your side into a single commit.

delba commented:
Please use `git rebase -i HEAD~19`, then `fixup` the commits into one (and maybe `reword` it) and finally `git push -f`

It will be more convenient to browse the git history. Thanks 😃

basememara committed:
Added support for NSDate parsing.

basememara commented:
Ok git history should be cleaned up now. Let me know if there's anything else you'd like me to do. Thanks for the tips again.

delba commented:
That's a great addition to JASON!!

Thank you so much Basem 😃

basememara commented:
Awesome, thank you!! 🎉

delba commented:
@basememara The `nsDate` property is available in `2.2`. Thanks again 😃

basememara commented:
Sweet!! Glad I could contribute to such a great library 👍
